pkg/ts/query.go

```diff
@@ -460,26 +469,40 @@ func (ii *interpolatingIterator) midTimestamp() int64 {
 	return dsi.underlyingData.startNanos + (int64(ii.offset) * dsi.sampleNanos) + (dsi.sampleNanos / 2)
 }
 
-// value returns the value at the current offset of this iterator.
+// value returns the value at the current offset of this iterator, or the
+// derivative at the current offset.
 func (ii *interpolatingIterator) value() float64 {
 	if !ii.isValid() {
 		return 0
 	}
-	if ii.nextReal.offset() == ii.offset {
+	isDerivative := ii.derivative != tspb.TimeSeriesQueryDerivative_NONE
+	if !isDerivative && ii.nextReal.offset() == ii.offset {
 		return ii.nextReal.value()
 	}
-	// Cannot interpolate if previous value is invalid.
+	// Cannot interpolate or compute derivative if previous value is invalid.
 	if !ii.prevReal.isValid() {
 		return 0
 	}
 
-	// Linear interpolation of value at the current offset.
+	// Linear interpolation of derivative or value at the current offset.
 	off := float64(ii.offset)
-	nextAvg := ii.nextReal.value()
+	nextVal := ii.nextReal.value()
 	nextOff := float64(ii.nextReal.offset())
-	prevAvg := ii.prevReal.value()
+	prevVal := ii.prevReal.value()
 	prevOff := float64(ii.prevReal.offset())
-	return prevAvg + (nextAvg-prevAvg)*(off-prevOff)/(nextOff-prevOff)
+
+	// Note: The derivative value could be factored out into a variable and used
+	// for all of these returns, but doing so seems to introduce a greater
+	// incidence of floating point artifacts.
```

Comment:
Ok. Probably something to do with the order of operations. If you grouped the `isDerivative` case as `(off-prevOff)*((nextVal-prevVal)/(nextOff-prevOff))` you might see the same behavior.